text/0000-extern-existential-type.md

```diff
@@ -0,0 +1,217 @@
+    }
+}
+
+impl existential type alloc::Heap = JemallocHeap;
+```
+
+By defining traits for each bit of deferred functionary, we can likewise cover each of the other use-cases.
+This frees the compiler and programmer to forget about the specific instances and just learn the general pattern.
+This is especially important for `log`, which isn't a sysroot crate and thus isn't known to the compiler at all at the moment.
+It would be very hard to justify special casing it with e.g. another attribute as the problem is solved today, when it needs none at the moment.
+As for the cost concerns with the existing techniques, No code is generated until the `impl existential type` is created, similar to with generic types, so there is no run-time cost whatsoever.
+
+Many of the mechanisms I listed above are on the verge of stabilization.
+I don't want to appear to by tying things up forever, so the design I've picked strives to be simple
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+It's best to understand this feature in terms of regular `existential type`.
+Type checking when *using* `pub extern existential type` works exactly the same way:
+The type is opaque except for it's bounds, and no traits can be implemented for it.
+```rust
+pub extern existential type Foo: Baz;
+existential type Bar: Baz;
+// both are used the same way if we
+```
+C and C++ programmers too will be familiar with the remote definition aspect of this from those language's "forward declarations" and their use in header files.
+
+On the definition side, since it is explicitly defined (somewhere), there are no inference constraints on items in the same module as the declaration or definition.
+
+One more interesting difference is the scope of where the type is transparent vs opaque: i.e. where can we see the type's definition, or only it's bounds.
+Just as in C where one gets:
+```rust
+struct Foo;
+
+// I know nothing about Foo
+
+struct Foo { int a; };
+
+// Ah now I do
+```
+when the `impl existential type` is in scope, the `existential existential type` can becomes transparent and behaves as if the declaration and definition were put together into a normal type alias.
+The definer can decide how gets to take advantage of it by making the definition public or not.
+```rust
+pub impl existential type alloc::Foo = Bar; // the big reveal
+impl existential type alloc::Foo = Bar; // the tiny reveal
+```
+There are no restrictions on the type of publicity compared to other items.
+
+Only one crate in the build plan can define the `pub extern existential type`.
+Unlike the trait system, there are no orphan restrictions that ensure crates can always be composed:
+any crate is free to define the `pub extern existential type`, as long is it isn't used with another that also does, in which case the violation will only be caught when building a crate that depends on both (or if one of the crates depends on the other).
```

Comment:
I think people will instinctively object to this. Talk more about why no orphan restrictions are OK here.